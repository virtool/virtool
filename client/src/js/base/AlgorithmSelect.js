import React from "react";
import PropTypes from "prop-types";
import { map } from "lodash-es";

import { Input } from "./";
import { getTaskDisplayName } from "../utils";

export const algorithms = [
    "pathoscope_bowtie",
    "nuvs"
];

/**
 * An input-based component for selecting an algorithm (eg. pathoscope_bowtie), by its display name
 * (eg. Pathoscope Bowtie).
 *
 */
export const AlgorithmSelect = (props) => (
    <Input
        name="algorithm"
        type="select"
        label={props.noLabel ? null : "Algorithm"}
        value={props.value}
        onChange={props.onChange}
    >
        {map(algorithms, algorithm =>
            <option
                key={algorithm}
                value={algorithm}
                disabled={(algorithm === "nuvs" && !props.hasHmm)}
            >
                {getTaskDisplayName(algorithm)}
            </option>
        )}
    </Input>
);

AlgorithmSelect.propTypes = {
    noLabel: PropTypes.bool,
    value: PropTypes.oneOf(["pathoscope_bowtie", "nuvs"]),
    onChange: PropTypes.func,
    hasHmm: PropTypes.bool
};

