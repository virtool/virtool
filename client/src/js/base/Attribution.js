import { capitalize } from "lodash-es";
import React from "react";
import styled from "styled-components";
import { InitialIcon } from "./InitialIcon";
import { RelativeTime } from "./RelativeTime";

export const UnstyledAttribution = ({ className, time, user, verb = "created" }) => {
    return (
        <span className={className}>
            {user ? <InitialIcon size="md" handle={user} /> : null}
            <span>{user}</span>
            <span>{user ? verb : capitalize(verb)}</span>
            <RelativeTime time={time} />
        </span>
    );
};

export const Attribution = styled(UnstyledAttribution)`
    align-items: center;
    display: inline-flex;
    font-size: inherit;

    *:not(:first-child) {
        margin-left: 3px;
    }
    .InitialIcon {
        margin-right: 5px;
    }
`;
