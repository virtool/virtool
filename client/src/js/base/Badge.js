import styled from "styled-components";

export const Badge = styled.span`
    background-color: #777777;
    border-radius: 10px;
    display: inline-block;
    min-width: 10px;
    padding: 3px 7px;
    font-size: 12px;
    font-weight: bold;
    color: #ffffff;
    line-height: 1;
    vertical-align: middle;
    white-space: nowrap;
    text-align: center;
`;

export const RedBadge = styled(Badge)`
    background-color: #c53030;
    color: #ffffff;
`;
