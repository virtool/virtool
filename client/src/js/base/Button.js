import { get } from "lodash-es";
import PropTypes from "prop-types";
import React from "react";
import { NavLink } from "react-router-dom";
import styled from "styled-components";
import { Icon } from "./Icon";
import { Tooltip } from "./Tooltip";

const getButtonBackgroundColor = ({ active, color, theme }) => {
    if (!color || color === "grey") {
        return active ? theme.color.greyLightest : theme.color.white;
    }

    return get(theme, ["color", `${color}${active ? "Dark" : ""}`], theme.white);
};

const getButtonBorderColor = ({ color, theme }) => {
    if (!color || color === "grey") {
        return theme.color.greyLight;
    }

    return theme.color[color];
};

const getButtonForegroundColor = ({ color, theme }) => {
    if (!color || color === "grey") {
        return theme.color.black;
    }

    return theme.color.white;
};

const getButtonHoverColor = ({ color, theme }) => {
    if (color && color !== "grey") {
        return get(theme, ["color", `${color}Dark`], theme.white);
    }

    return theme.color.greyLightest;
};

export const StyledButton = styled.button`
    align-items: center;
    background-color: ${getButtonBackgroundColor};
    background-image: none;
    border: 1px solid ${getButtonBorderColor};
    border-radius: ${props => props.theme.borderRadius.sm};
    box-shadow: ${props => (props.active ? props.theme.boxShadow.inset : "")};
    color: ${getButtonForegroundColor};
    cursor: pointer;
    display: inline-flex;
    font-weight: 500;
    justify-content: center;
    margin-bottom: 0;
    min-width: 42px;
    min-height: 36px;
    opacity: ${props => (props.disabled ? 0.5 : 1)};
    padding: 0 10px;
    position: relative;
    user-select: none;
    touch-action: manipulation;
    transition: box-shadow 200ms ease-in-out;
    white-space: nowrap;

    i + span {
        margin-left: 5px;
    }

    :focus {
        color: ${getButtonForegroundColor};
    }

    :disabled {
        cursor: not-allowed;
    }

    :not(:disabled):hover {
        background-color: ${getButtonHoverColor};
        border-color: ${getButtonHoverColor};
        color: ${getButtonForegroundColor};
    }

    :not(:disabled)::after {
        border-radius: ${props => props.theme.borderRadius.sm};
        box-shadow: ${props => props.theme.boxShadow.lg};
        content: "";
        height: 100%;
        width: 100%;
        opacity: 0;
        position: absolute;
        transition: opacity 150ms ease, scale 150ms ease;
    }

    :not(:disabled):hover::after {
        opacity: 1;
    }
`;

export const LinkButton = ({ children, color, className, icon, replace, tip, to }) => {
    const button = (
        <StyledButton as={NavLink} className={className} color={color} replace={replace} to={to}>
            {icon && <Icon name={icon} />}
            {children ? <span>{children}</span> : null}
        </StyledButton>
    );

    if (tip) {
        return <Tooltip tip={tip}>{button}</Tooltip>;
    }

    return button;
};

LinkButton.propTypes = {
    children: PropTypes.node,
    color: PropTypes.string,
    className: PropTypes.string,
    icon: PropTypes.string,
    replace: PropTypes.bool,
    tip: PropTypes.string,
    tipPlacement: PropTypes.oneOf(["top", "right", "bottom", "left"]),
    to: PropTypes.oneOfType([PropTypes.object, PropTypes.string]).isRequired
};

LinkButton.defaultProps = {
    tipPlacement: "top"
};

export const Button = ({ active, children, className, color, disabled, icon, tip, type, onBlur, onClick }) => {
    const button = (
        <StyledButton
            active={active}
            className={className}
            color={color}
            disabled={disabled}
            type={type}
            onBlur={onBlur}
            onClick={onClick}
        >
            {icon && <Icon name={icon} />}
            {children ? <span>{children}</span> : null}
        </StyledButton>
    );

    if (tip) {
        return <Tooltip tip={tip}>{button}</Tooltip>;
    }

    return button;
};

Button.propTypes = {
    active: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    color: PropTypes.string,
    disabled: PropTypes.bool,
    icon: PropTypes.string,
    tip: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    tipPlacement: PropTypes.oneOf(["top", "right", "bottom", "left"]),
    type: PropTypes.oneOf(["button", "submit"]),
    onBlur: PropTypes.func,
    onClick: PropTypes.func
};

Button.defaultProps = {
    color: "grey",
    disabled: false,
    tipPlacement: "top",
    type: "button"
};
