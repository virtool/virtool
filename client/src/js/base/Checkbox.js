import CX from "classnames";
import React from "react";
import PropTypes from "prop-types";

/**
 * Application icon font.
 *
 * @param checked
 * @param partial
 * @returns {*}
 * @constructor
 */
export const CheckboxIcon = ({ checked, partial }) => {

    let name;

    if (checked) {
        name = "checked";
    } else if (partial) {
        name = "partial";
    } else {
        name = "unchecked";
    }

    return <i className={`i-checkbox-${name}`} />;
};


/**
 * A simple checkbox component based on the application icon font.
 *
 * @param props
 * @returns {*} any type
 * @constructor
 */
export const Checkbox = (props) => {

    let className = CX("pointer", {
        "pull-right": props.pullRight,
        "labelled-checkbox": props.label,
        "text-muted": props.disabled
    });

    if (props.className) {
        className += ` ${props.className}`;
    }

    let style = {cursor: props.disabled ? "not-allowed" : "pointer"};

    if (props.style) {
        style = {...style, ...props.style};
    }

    return (
        <span className={className} onClick={props.disabled ? null : props.onClick} style={style}>
            <CheckboxIcon {...props} /> {props.label ? <span>{props.label}</span> : null}
        </span>
    );
};

Checkbox.propTypes = {
    checked: PropTypes.bool,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    label: PropTypes.node,
    onClick: PropTypes.func,
    partial: PropTypes.bool,
    pullRight: PropTypes.bool,
    style: PropTypes.object
};

Checkbox.defaultProps = {
    checked: false,
    partial: false,
    pullRight: false
};
