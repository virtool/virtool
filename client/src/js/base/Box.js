import { Link } from "react-router-dom";
import styled from "styled-components";
import { getBorder, getFontSize, getFontWeight } from "../app/theme";
import { Badge } from "./Badge";
import { CheckboxLabel, StyledCheckbox } from "./Checkbox";
import { Table } from "./Table";

export const Box = styled.div`
    border: ${getBorder};
    border-radius: ${props => props.theme.borderRadius.sm};
    box-sizing: border-box;
    cursor: ${props => (props.onClick ? "pointer" : "auto")};
    margin-bottom: 15px;
    padding: 10px 15px;
    position: relative;

    &:hover {
        ${props => (props.onClick ? `background-color: ${props.theme.color.greyHover};` : "")}
    }
`;

export const BoxGroup = styled(Box)`
    border-radius: ${props => props.theme.borderRadius.sm};
    padding: 0;
    position: relative;

    & > ${Table} {
        border: none;
        margin: 0;

        &:first-child {
            border-top: none;
        }

        tbody {
            border-top: none;
        }

        td,
        th {
            min-width: 0;
            padding: 8px 15px;
        }
    }
`;

export const BoxGroupSection = styled.div`
    background-color: transparent;
    border: none;
    border-radius: 0;
    color: inherit;
    cursor: ${props => (props.onClick && !props.active ? "pointer" : "auto")};
    padding: 10px 15px;
    position: relative;

    &[disabled] {
        background-color: ${props => props.theme.color.greyHover};
        cursor: not-allowed;
        color: ${props => props.theme.color.grey};
        user-select: none;
    }

    &:hover {
        ${props => (props.onClick && !props.active ? `background-color: ${props.theme.color.greyHover};` : "")}
    }

    &:not(:last-child) {
        border-bottom: ${getBorder};
    }
`;

export const SelectBoxGroupSection = styled(BoxGroupSection)`
    background-color: ${props => (props.active ? props.theme.color.blue : "transparent")};
    color: ${props => (props.active ? props.theme.color.white : "inherit")};
    cursor: pointer;
    width: 100%;

    &:focus {
        box-shadow: inset 0 0 0 2px rgba(43, 108, 176, 0.5);
        outline: none;
    }

    ${StyledCheckbox} {
        background-color: ${props => (props.active ? props.theme.color.white : "transparent")};
        color: ${props => props.theme.color[props.active ? "blueDark" : "greyLight"]};
        margin-right: 10px;
    }

    ${CheckboxLabel} {
        margin: 0;
    }
`;

export const BoxGroupHeader = styled(BoxGroupSection)`
    align-items: stretch;
    background-color: ${props => props.theme.color.greyLightest};
    display: flex;
    flex-direction: column;
    font-size: ${getFontSize("md")};
    padding: 15px 15px 12px;

    h2 {
        align-items: center;
        display: flex;
        font-size: ${props => props.theme.fontSize.lg};
        font-weight: ${getFontWeight("thick")};
        margin: 0;

        ${Badge} {
            margin-left: 5px;
        }
    }

    p {
        color: ${props => props.theme.color.greyDarkest};
        margin: 5px 0 0;
    }
`;

export const BoxTitle = styled.h1`
    font-size: ${props => props.theme.fontSize.md};
    font-weight: ${props => props.theme.fontWeight.thick};
    margin: 5px 0 15px 0;
`;

export const SpacedBox = styled(Box)`
    box-shadow: ${props => props.theme.boxShadow.md};
    margin-bottom: 10px;
`;

export const LinkBox = styled(Link)`
    border: ${getBorder};
    border-radius: ${props => props.theme.borderRadius.sm};
    box-shadow: ${props => props.theme.boxShadow.sm};
    box-sizing: border-box;
    color: ${props => props.theme.color.black};
    cursor: pointer;
    display: block;
    margin-bottom: 12px;
    padding: 10px 15px;
    position: relative;

    &:hover {
        background-color: ${props => props.theme.color.greyHover};
    }
`;
