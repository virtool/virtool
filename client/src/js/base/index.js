import { AlgorithmSelect } from "./AlgorithmSelect";
import { AutoProgressBar, ProgressBar } from "./ProgressBar";
import { bsStyles } from "./utils";
import { Checkbox } from "./Checkbox";
import { Flex, FlexItem } from "./Flex";
import { Help } from "./Help";
import { Icon } from "./Icon";
import { Identicon } from "./Identicon";
import { Input } from "./Input";
import { InputError } from "./InputError";
import { InputSave } from "./InputSave";
import { ListGroupItem } from "./ListGroupItem";
import { NotFound } from "./NotFound";
import { Panel } from "./Panel";
import { Radio } from "./Radio";
import { RelativeTime } from "./RelativeTime";
import { RemoveBanner } from "./RemoveBanner";
import { RemoveModal } from "./RemoveModal";
import { SaveButton } from "./SaveButton";
import { ScrollList } from "./ScrollList";
import { Tooltip } from "./Tooltip";
import { UploadBar } from "./UploadBar";
import { ViewHeader } from "./ViewHeader";

export * from "./Alert";
export * from "./Attribution";
export * from "./Badge";
export * from "./Box";
export * from "./Breadcrumb";
export * from "./Breakpoints";
export * from "./Button";
export * from "./ButtonGroup";
export * from "./ButtonToolbar";
export * from "./Code";
export * from "./Container";
export * from "./ExternalLink";
export * from "./Label";
export * from "./Loader";
export * from "./LoadingPlaceholder";
export * from "./Logo";
export * from "./Modal";
export * from "./NoneFound";
export * from "./Table";
export * from "./SectionHeader";
export * from "./SlashList";
export * from "./TabLink";
export * from "./Tabs";
export * from "./Toolbar";
export * from "./ViewHeader";
export * from "./SelectBox";

export {
    AlgorithmSelect,
    AutoProgressBar,
    bsStyles,
    Checkbox,
    Flex,
    FlexItem,
    Help,
    Icon,
    Identicon,
    Input,
    InputError,
    InputSave,
    ListGroupItem,
    NotFound,
    Panel,
    ProgressBar,
    Radio,
    RelativeTime,
    RemoveBanner,
    RemoveModal,
    SaveButton,
    ScrollList,
    Tooltip,
    UploadBar,
    ViewHeader
};
