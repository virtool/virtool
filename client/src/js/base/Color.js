import React, { useCallback } from "react";
import styled from "styled-components";
import { borderRadius } from "../app/theme";
import { Box } from "./Box";
import { Input } from "./Input";

const colors = [
    // Grey
    "D1D5DB",
    "6B7280",
    "374151",

    // Red
    "FCA5A5",
    "EF4444",
    "B91C1C",

    // Yellow
    "FCD34D",
    "F59E0B",
    "B45309",

    // Green
    "6EE7B7",
    "10B981",
    "047857",

    // Blue
    "93C5FD",
    "3B82F6",
    "1D4ED8",

    // Indigo
    "A5B4FC",
    "6366F1",
    "4338CA",

    // Purple
    "C4B5FD",
    "8B5CF6",
    "5B21B6",

    // Pink
    "FBCFE8",
    "F472B6",
    "EC4899"
];

const StyledColor = styled(Box)`
    padding: 10px;
`;

const UnstyledColorSquare = ({ className, color, onClick }) => {
    const handleClick = useCallback(() => onClick(color), [color, onClick]);
    return <button className={className} type="button" title={color} onClick={handleClick} />;
};

const ColorSquare = styled(UnstyledColorSquare)`
    backface-visibility: hidden;
    background-color: ${props => props.color};
    background-image: none;
    border: none;
    cursor: pointer;
    flex: 1 0 auto;
    width: auto;

    &:first-child {
        border-bottom-left-radius: ${borderRadius.sm};
        border-top-left-radius: ${borderRadius.sm};
    }

    &:last-child {
        border-bottom-right-radius: ${borderRadius.sm};
        border-top-right-radius: ${borderRadius.sm};
    }

    &:focus {
        box-shadow: rgb(0 0 0 / 25%) 0 0 5px 2px;
        outline: solid white 2px;
        z-index: 2;
    }

    &:hover {
        transform: translateY(-2px);
    }
`;

const ColorGrid = styled.div`
    align-items: stretch;
    background-color: transparent;
    display: flex;
    height: 30px;
    margin-top: 10px;
`;

export const Color = ({ id, value, onChange }) => {
    const squares = colors.map(color => <ColorSquare key={color} color={`#${color}`} onClick={onChange} />);

    return (
        <StyledColor>
            <Input id={id} value={value} onChange={e => onChange(e.target.value)} />
            <ColorGrid>{squares}</ColorGrid>
        </StyledColor>
    );
};
