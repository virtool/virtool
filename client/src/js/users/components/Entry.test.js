import UserEntry from "./Entry";
import { Icon } from "../../base";

describe("<UserEntry />", () => {

    describe("renders correctly", () => {
        let initialState;
        let store;
        let wrapper;

        it("with admin icon when [isAdmin=true]", () => {
            initialState = {
                users: {
                    list: {
                        documents: [
                            {
                                id: "testid",
                                identicon: "randomhashof15c",
                                administrator: true
                            }
                        ]
                    }
                }
            };
            store = mockStore(initialState);
            wrapper = shallow(<UserEntry store={store} index={0} />).dive();

            expect(wrapper.find(Icon).length).toBe(1);
            expect(wrapper).toMatchSnapshot();
        });

        it("without icon otherwise", () => {
            initialState = {
                users: {
                    list: {
                        documents: [
                            {
                                id: "testid",
                                identicon: "randomhashof15c",
                                administrator: false
                            }
                        ]
                    }
                }
            };
            store = mockStore(initialState);
            wrapper = shallow(<UserEntry store={store} index={0} />).dive();

            expect(wrapper.find(Icon).length).toBe(0);
            expect(wrapper).toMatchSnapshot();
        });

    });

});
