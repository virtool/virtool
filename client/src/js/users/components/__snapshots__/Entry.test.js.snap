// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserEntry /> renders correctly with admin icon when [isAdmin=true] 1`] = `
<LinkContainer
  activeClassName="active"
  exact={false}
  replace={false}
  strict={false}
  style={
    Object {
      "paddingLeft": "10px",
    }
  }
  to="/administration/users/testid"
>
  <ListGroupItem
    bsClass="list-group-item"
    className="spaced"
    listItem={false}
  >
    <Flex
      alignContent="stretch"
      alignItems="center"
      direction="row"
      justifyContent="flex-start"
      wrap="nowrap"
    >
      <Identicon
        hash="randomhashof15c"
        size={32}
      />
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      >
        testid
      </FlexItem>
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      >
        <Icon
          bsStyle="primary"
          faStyle="fas"
          fixedWidth={false}
          name="user-shield"
          pullRight={false}
        />
      </FlexItem>
    </Flex>
  </ListGroupItem>
</LinkContainer>
`;

exports[`<UserEntry /> renders correctly without icon otherwise 1`] = `
<LinkContainer
  activeClassName="active"
  exact={false}
  replace={false}
  strict={false}
  style={
    Object {
      "paddingLeft": "10px",
    }
  }
  to="/administration/users/testid"
>
  <ListGroupItem
    bsClass="list-group-item"
    className="spaced"
    listItem={false}
  >
    <Flex
      alignContent="stretch"
      alignItems="center"
      direction="row"
      justifyContent="flex-start"
      wrap="nowrap"
    >
      <Identicon
        hash="randomhashof15c"
        size={32}
      />
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      >
        testid
      </FlexItem>
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      />
    </Flex>
  </ListGroupItem>
</LinkContainer>
`;
